typing/rec_check.ml

```diff
@@ -306,338 +322,671 @@ let classify_expression : Typedtree.expression -> sd =
+        expression e << Dereference;
+      ]
+    | Texp_letexception ({ext_id}, e) ->
+      (* G |- e: m
+         ----------------------------
+         G |- let exception A in e: m
+      *)
+      remove_id ext_id (expression e)
     | Texp_assert e ->
-        Use.inspect (expression env e)
-    | Texp_pack m ->
-        modexp env m
+      (*
+        G |- e: m[Dereference]
+        -----------------------
+        G |- assert e: m
+
+        Note: `assert e` is treated just as if `assert` was a function.
+      *)
+      expression e << Dereference
+    | Texp_pack mexp ->
+      (*
+        G |- M: m
+        ----------------
+        G |- module M: m
+      *)
+      modexp mexp
     | Texp_object (clsstrct, _) ->
-        class_structure env clsstrct
+      class_structure clsstrct
     | Texp_try (e, cases) ->
-      (* This is more permissive than the old check. *)
-      let case env {Typedtree.c_rhs} = expression env c_rhs in
-      Use.join (expression env e)
-        (list case env cases)
-    | Texp_override (_, fields) ->
-      let field env (_, _, e) = expression env e in
-      Use.inspect (list field env fields)
+      (*
+        G |- e: m      (Gi; _ |- pi -> ei : m)^i
+        --------------------------------------------
+        G + sum(Gi)^i |- try e with (pi -> ei)^i : m
+
+        Contrarily to match, the patterns p do not inspect
+        the value of e, so their mode does not influence the
+        mode of e.
+      *)
+      let case_env c m = fst (case c m) in
+      join [
+        expression e;
+        list case_env cases;
+      ]
+    | Texp_override (pth, fields) ->
+      (*
+         G |- pth : m   (Gi |- ei : m[Derefence])^i
+         ----------------------------------------------------
+         G + sum(Gi)^i |- {< (xi = ei)^i >} (at path pth) : m
+
+         Note: {< .. >} is desugared to a function application, but
+         the function implementation might still use its arguments in
+         a guarded way only -- intuitively it should behave as a constructor.
+         We could possibly refine the arguments' Dereference into Guard here.
+      *)
+      let field (_, _, arg) = expression arg in
+      join [
+        path pth << Dereference;
+        list field fields << Dereference;
+      ]
     | Texp_function { cases } ->
-      Use.delay (list (case ~scrutinee:Use.empty) env cases)
+      (*
+         (Gi; _ |- pi -> ei : m[Delay])^i
+         --------------------------------------
+         sum(Gi)^i |- function (pi -> ei)^i : m
+
+         Contrarily to match, the value that is pattern-matched
+         is bound locally, so the pattern modes do not influence
+         the final environment.
+      *)
+      let case_env c m = fst (case c m) in
+      list case_env cases << Delay
     | Texp_lazy e ->
-        begin match Typeopt.classify_lazy_argument e with
+      (*
+        G |- e: m[Delay]
+        ----------------  (modulo some subtle compiler optimizations)
```

Comment:
I'm expecting this to break when somebody implements (e.g.) the lazy value optimization for unboxed constructors.  (But perhaps having the common code in `Typeopt` will be enough to prevent that.)

Reply:
Yes, I'm expecting such future changes to return one of the known-to-be-shortcut variants (for unboxed constructors, typically the classifier of the underlying argument), or to add a new variants to be handled properly here.